GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs

```diff
@@ -1078,30 +1087,85 @@ private void UpdatePlaceholders()
+                {
+                    updatedPlaceholderDictionary = new ConcurrentDictionary<string, PlaceholderListDatabase.PlaceholderData>(
+                        concurrencyLevel: numThreads,
+                        capacity: placeholderFilesListCopy.Count + placeholderFoldersListCopy.Count,
+                        comparer: StringComparer.Ordinal);
+                    updatedPlaceholderBag = null;
+                    addPlaceholderToUpdatedPlaceholders = (data) => updatedPlaceholderDictionary.TryAdd(data.Path, data);
+                }
+                else
+                {
+                    updatedPlaceholderDictionary = null;
+                    updatedPlaceholderBag = new ConcurrentBag<PlaceholderListDatabase.PlaceholderData>();
+                    addPlaceholderToUpdatedPlaceholders = (data) => updatedPlaceholderBag.Add(data);
+                }
+                
                 this.ProcessListOnThreads(
-                    placeholderListCopy.Where(x => !x.IsFolder).ToList(),
+                    numThreads,
+                    placeholderFilesListCopy,
                     (placeholderBatch, start, end, blobSizesConnection, availableSizes) => 
                         this.BatchPopulateMissingSizesFromRemote(blobSizesConnection, placeholderBatch, start, end, availableSizes),
                     (placeholder, blobSizesConnection, availableSizes) => 
-                        this.UpdateOrDeleteFilePlaceholder(blobSizesConnection, placeholder, updatedPlaceholderList, folderPlaceholdersToKeep, availableSizes));
+                        this.UpdateOrDeleteFilePlaceholder(blobSizesConnection, placeholder, addPlaceholderToUpdatedPlaceholders, folderPlaceholdersToKeep, availableSizes));
 
                 this.blobSizes.Flush();
 
-                // Waiting for the UpdateOrDeleteFilePlaceholder to fill the folderPlaceholdersToKeep before trying to remove folder placeholders
-                // so that we don't try to delete a folder placeholder that has file placeholders and just fails
-                foreach (PlaceholderListDatabase.PlaceholderData folderPlaceholder in placeholderListCopy.Where(x => x.IsFolder).OrderByDescending(x => x.Path))
+                using (BlobSizes.BlobSizesConnection blobSizesConnection = this.blobSizes.CreateConnection())
+                {
+                    // A hash of the folder placeholders is only required if the platform expands directories
+                    HashSet<string> folderPlaceholders = 
+                        GVFSPlatform.Instance.KernelDriver.EnumerationExpandsDirectories ?
+                                    new HashSet<string>(placeholderFoldersListCopy.Select(x => x.Path), StringComparer.OrdinalIgnoreCase) : 
+                                    null;
+                    
+                    foreach (PlaceholderListDatabase.PlaceholderData folderPlaceholder in placeholderFoldersListCopy.OrderByDescending(x => x.Path))

```

Comment:
Is this `OrderByDescending` here to ensure that we walk through the tree from the bottom up? I would add some comments to clarify that.